couchdb/server/helpers.py

```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+#
+from inspect import getsource
+from textwrap import dedent
+from types import FunctionType
+
+from couchdb import util
+
+
+def maybe_extract_source(fun):
+    if isinstance(fun, FunctionType):
+        return dedent(getsource(fun))
+    elif isinstance(fun, util.strbase):
+        return fun
+    raise TypeError('Function object or source string expected, got %r' % fun)
```

Comment:
Err... It seems that it blew up when user put some builtin function here.

e.g.:
```python
In [20]: sorted
Out[20]: <function sorted>

In [21]: type(sorted)
Out[21]: builtin_function_or_method

In [22]: inspect.getsource(sorted)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-22-c9deb9140b13> in <module>()
----> 1 inspect.getsource(sorted)

/home/iblis/py/py36/lib/python3.6/inspect.py in getsource(object)
    925     or code object.  The source code is returned as a single string.  An
    926     OSError is raised if the source code cannot be retrieved."""
--> 927     lines, lnum = getsourcelines(object)
    928     return ''.join(lines)
    929 

/home/iblis/py/py36/lib/python3.6/inspect.py in getsourcelines(object)
    912     raised if the source code cannot be retrieved."""
    913     object = unwrap(object)
--> 914     lines, lnum = findsource(object)
    915 
    916     if ismodule(object):

/home/iblis/py/py36/lib/python3.6/inspect.py in findsource(object)
    725     is raised if the source code cannot be retrieved."""
    726 
--> 727     file = getsourcefile(object)
    728     if file:
    729         # Invalidate cache if needed.

/home/iblis/py/py36/lib/python3.6/inspect.py in getsourcefile(object)
    641     Return None if no way can be identified to get the source.
    642     """
--> 643     filename = getfile(object)
    644     all_bytecode_suffixes = importlib.machinery.DEBUG_BYTECODE_SUFFIXES[:]
    645     all_bytecode_suffixes += importlib.machinery.OPTIMIZED_BYTECODE_SUFFIXES[:]

/home/iblis/py/py36/lib/python3.6/inspect.py in getfile(object)
    623         return object.co_filename
    624     raise TypeError('{!r} is not a module, class, method, '
--> 625                     'function, traceback, frame, or code object'.format(object))
    626 
    627 def getmodulename(path):

TypeError: <built-in function sorted> is not a module, class, method, function, traceback, frame, or code object
```